### PR TITLE
Use the instance class as the logging category

### PR DESCRIPTION
**Pull request by ehuelsmann**
# Description

Per Log::Log4perl's recommendation (see https://metacpan.org/pod/Log::Log4perl#Pitfalls-with-Categories),
use the class of the instance as the category for purposes of logging -- where possible.

This allows much better control over which parts of the application do
logging and which parts don't.

## Type of change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

The reason I'm considering this a breaking change is because logging configuration might need to be adjusted for *inherited classes* (not for Workflow classes themselves): Since I don't expect there to be a specific logging configuration for the `Workflow` category, all the loggers will fall back to the current default, which they already did by calling `get_logger()`. For any inherited loggers in different toplevel namespaces (e.g. `LedgerSMB::Workflow::Action::Email`), a different logging category may be selected (i.e. `LedgerSMB`), resulting in different logging results than before.

Where to document this?

# Checklist:

- [x] My code follows the style guidelines of this project, please see the [contribution guidelines](https://github.com/jonasbn/perl-workflow/blob/master/CONTRIBUTING.md).
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ehuelsmann** [comment]
Travis says I made a mistake here. I'll fix that before you should look at it.

**Comment by ehuelsmann** [comment]
> Travis says I made a mistake here. I'll fix that before you should look at it.

Fixed.

**Comment by ehuelsmann** [comment]
@jonasbn this branch is ready to be merged. The reason it says checks are failing is because coverage dropped by 0.1%, which is a consequence of me having removed more tested than untested code lines.
However, I don't see how I can improve coverage while sticking to the intent of the PR.

Please review.

**Comment by jonasbn** [comment]
Hi @ehuelsmann 

Even though it looks massive, the theme of the PR is somewhat uniform. I do not expect this PR review to be as extensive. I will see if I can find the time shortly.

**Comment by ehuelsmann** [comment]
@jonasbn no hurry; this one is indeed much simpler than the other one. If I can choose, I also prefer this one, because it touches more code and would be nice to have that out of the way for any further PRs. The other one is much more localized.

File: lib/Workflow/Base.pm [hunk 1]
```diff
@@ -20,7 +20,11 @@ sub new {
     return $self;
 }
 
-sub init {return}
+sub init {return};
+
+sub log {
+    return ($_[0]->{log} ||=  Log::Log4perl->get_logger(ref $_[0]) );
```
Comment (jonasbn) [inline lib/Workflow/Base.pm:26]: Cosmetic change required, the parentheses should use same spacing, so it does not look as something is missing.

```suggestion
return ( $_[0]->{log} ||=  Log::Log4perl->get_logger(ref $_[0]) );
````

File: lib/Workflow/Persister/DBI/AutoGeneratedId.pm [hunk 2]
```diff
@@ -13,7 +13,6 @@ __PACKAGE__->mk_accessors(@FIELDS);
 
 sub new {
     my ( $class, $params ) = @_;
-    my $log = get_logger();
```
Comment (jonasbn) [inline lib/Workflow/Persister/DBI/AutoGeneratedId.pm:16]: With the removal of this statement, I cannot see where the log handle comes from in this class, it does not inherit from any classes provided the log handle via the main object - or am I missing something?

File: lib/Workflow/Persister/DBI/SequenceId.pm [hunk 3]
```diff
@@ -13,14 +13,12 @@ $Workflow::Persister::DBI::SequenceId::VERSION = '1.52';
 my @FIELDS = qw( sequence_name sequence_select );
 __PACKAGE__->mk_accessors(@FIELDS);
 
-my ($log);
```
Comment (jonasbn) [inline lib/Workflow/Persister/DBI/SequenceId.pm:16]: With the removal of this statement, I cannot see where the log handle comes from in this class, it does not inherit from any classes provided the log handle via the main object - or am I missing something?

**Review by jonasbn — CHANGES_REQUESTED** [review]
Hi @ehuelsmann 

I have made some minor remarks. Could you please go over them. One is cosmetic and two are actually the same pattern, where I might just be missing something.

File: lib/Workflow/Persister/DBI/AutoGeneratedId.pm (hunk 2 above)
Comment (ehuelsmann) [reply lib/Workflow/Persister/DBI/AutoGeneratedId.pm:16]: You're totally right. Adding the `log` attribute to both classes and further aligning the two (one has a "new" method, the other one doesn't...)

Comment (ehuelsmann) [reply lib/Workflow/Persister/DBI/AutoGeneratedId.pm:16]: I've replaced this now with a `log` accessor in the class which is initialized upon instantiation.

File: lib/Workflow/Persister/DBI/SequenceId.pm (hunk 3 above)
Comment (ehuelsmann) [reply lib/Workflow/Persister/DBI/SequenceId.pm:16]: As this is the same pattern, using the same solution. (I've replaced this now with a `log` accessor in the class which is initialized upon instantiation.)

**Review by jonasbn — APPROVED** [review]

**Comment by jonasbn** [comment]
Hi @ehuelsmann 

I think we should consider releasing this as a major release based on your initial comment and flagging as the PR as _breaking_

**Comment by ehuelsmann** [comment]
> Hi @ehuelsmann
> 
> I think we should consider releasing this as a major release based on your initial comment and flagging as the PR as _breaking_

Yes, I think that you're right. It's a more strict interpretation of SEMVER than I would have used on my own software, because I consider logging more of a secondary or supporting functionality -- I wouldn't have thought to apply SEMVER to that too (as the primary functionality and documented API didn't change).

Anyway, a question that comes up when you say that is: should we then plan for more rigorous change? Maybe changes including the API?

**Pull request merged**